doc/api/async_hooks.md

```diff
@@ -0,0 +1,670 @@
+
+```js
+class MyClass {
+  constructor() {
+    this._id = async_hooks.newId();
+    this._triggerId = async_hooks.initTriggerId();
+  }
+}
+```
+
+#### `async_hooks.initTriggerId()`
+
+* Returns {number}
+
+There are several ways to set the `triggerId` for an instantiated resource.
+This API is how that value is retrieved. It returns the `id` of the resource
+responsible for the newly created resource being instantiated. For example:
+
+
+#### `async_hooks.emitInit(id, type[, triggerId][, resource])`
+
+* `id` {number} Generated by calling `newId()`
+* `type` {String}
+* `triggerId` {number} **Default:** `currentId()`
+* `resource` {Object}
+* Returns {Undefined}
+
+Emit that a resource is being initialized. `id` should be a value returned by
+`async_hooks.newId()`. Usage will probably be as follows:
+
+```js
+class Foo {
+  constructor() {
+    this._id = async_hooks.newId();
+    this._triggerId = async_hooks.initTriggerId();
+    async_hooks.emitInit(this._id, 'Foo', this._triggerId, this);
+  }
+}
+```
+
+In the circumstance that the embedder needs to define a different trigger id
+than `currentId()`, they can pass in that id manually.
+
+It is suggested to have `emitInit()` be the last call in the object's
+constructor.
+
+
+#### `async_hooks.emitBefore(id[, triggerId])`
+
+* `id` {number} Generated by `newId()`
+* `triggerId` {number}
+* Returns {Undefined}
+
+Notify `before()` hooks that the resource is about to enter its execution call
+stack. If the `triggerId` of the resource is different from `id` then pass
+it in.
+
+Example usage:
+
+```js
+MyThing.prototype.done = function done() {
+  // First call the before() hooks. So currentId() shows the id of the
+  // resource wrapping the id that's been passed.
+  async_hooks.emitBefore(this._id);
+
+  // Run the callback.
+  this.callback();
+
+  // Call after() callbacks now that the old id has been restored.
+  async_hooks.emitAfter(this._id);
+};
+```
+
+#### `async_hooks.emitAfter(id)`
+
+* `id` {number} Generated by `newId()`
+* Returns {Undefined}
+
+Notify `after()` hooks that the resource is exiting its execution call stack.
+
+Even though the state of `id` is tracked internally, passing it in is required
+as a way to validate that the stack is unwinding properly.
+
+For example, no two async stack should cross when `emitAfter()` is called.
+
+```
+init # Foo
+init # Bar
+...
+before # Foo
+before # Bar
+after # Foo   <- Should be called after Bar
+after # Bar
+```
+
+**Note:** It is not necessary to wrap the callback in a try/finally and force
+emitAfter() if the callback throws. That is automatically handled by the
+fatal exception handler.
+
+#### `async_hooks.emitDestroy(id)`
+
+* `id` {number} Generated by `newId()`
+* Returns {Undefined}
+
+Notify hooks that a resource is being destroyed (or being moved to the free'd
+resource pool).
+
+#### `async_hooks.triggerIdScope(triggerId, callback)`
+
+* `triggerId` {number}
+* `callback` {Function}
+* Returns {Undefined}
```

Comment:
All types? Getting mixed info here. I was told elsewhere only primitive types.
For now I lowercased `Undefined` and `String` as requested above.

Going to wait for clarification here until I lower case more.